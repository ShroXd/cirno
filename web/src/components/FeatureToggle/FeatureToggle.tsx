import { ReactNode } from 'react'

import { useFeatureFlag } from '~/hooks/feature/useFeatureFlag'

interface FeatureToggleProps {
  featureId: string
  children?: ReactNode
  fallback?: ReactNode
}

export const FeatureToggle = ({
  featureId,
  children,
  fallback,
}: FeatureToggleProps) => {
  const isEnabled = useFeatureFlag(featureId)

  return isEnabled ? <>{children}</> : <>{fallback}</>
}
