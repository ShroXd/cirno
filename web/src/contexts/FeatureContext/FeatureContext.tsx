import { ReactNode, createContext, useEffect, useState } from 'react'

import { defaultFeatures } from '~/config/feature'

export interface Feature {
  id: string
  enabled: boolean
  name: string
  description: string
}

export interface FeatureContextType {
  features: Feature[]
  isFeatureEnabled: (featureId: string) => boolean
  toggleFeature: (featureId: string) => void
  updateFeature: (features: Feature[]) => void
}

export const FeatureContext = createContext<FeatureContextType>({
  features: [],
  isFeatureEnabled: () => false,
  toggleFeature: () => {},
  updateFeature: () => {},
})

export const FeatureProvider = ({ children }: { children: ReactNode }) => {
  const [features, setFeatures] = useState<Feature[]>(() => {
    const savedFeatures = localStorage.getItem('features')
    return savedFeatures ? JSON.parse(savedFeatures) : defaultFeatures
  })

  useEffect(() => {
    localStorage.setItem('features', JSON.stringify(features))
  }, [features])

  const isFeatureEnabled = (featureId: string) =>
    features.find(feature => feature.id === featureId)?.enabled ?? false

  const toggleFeature = (featureId: string) => {
    setFeatures(prevFeatures =>
      prevFeatures.map(feature =>
        feature.id === featureId
          ? { ...feature, enabled: !feature.enabled }
          : feature
      )
    )
  }

  const updateFeature = (features: Feature[]) => {
    setFeatures(features)
  }

  return (
    <FeatureContext.Provider
      value={{ features, isFeatureEnabled, toggleFeature, updateFeature }}
    >
      {children}
    </FeatureContext.Provider>
  )
}
