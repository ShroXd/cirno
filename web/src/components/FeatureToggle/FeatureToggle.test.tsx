import { render } from '@testing-library/react'
import { axe } from 'jest-axe'
import { describe, expect, it, vi } from 'vitest'

import { FeatureToggle } from './FeatureToggle'
import { FeatureProvider } from '~/contexts/FeatureContext/FeatureContext'

vi.mock('~/hooks/feature/useFeatureFlag', () => ({
  useFeatureFlag: vi.fn((id: string) => id === 'feature1'),
}))

describe('FeatureToggle', () => {
  it('should render without a11y violations', async () => {
    const { container } = render(<FeatureToggle featureId='feature1' />)
    expect(await axe(container)).toHaveNoViolations()
  })

  it('should render correctly with passed props', () => {
    const { queryByTestId, container } = render(
      <FeatureProvider>
        <FeatureToggle featureId='feature1'>
          <div data-testid='text-content-1'>Text content 1</div>
        </FeatureToggle>
        <FeatureToggle featureId='feature2'>
          <div data-testid='text-content-2'>Text content 2</div>
        </FeatureToggle>
      </FeatureProvider>
    )

    expect(queryByTestId('text-content-1')).toBeVisible()
    expect(queryByTestId('text-content-2')).toBeNull()

    expect(container).toMatchSnapshot()
  })
})
