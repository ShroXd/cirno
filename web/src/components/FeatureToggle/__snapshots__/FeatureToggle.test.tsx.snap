// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FeatureToggle > should render correctly with passed props 1`] = `
<div>
  <div>
    <div
      data-testid="text-content-1"
    >
      Text content 1
    </div>
  </div>
  <div />
</div>
`;
